scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -772,6 +781,135 @@ public final synchronized void destroy() {
+				return principal.equals(connectionPrincipal);
+			}
+		};
+		return startTerminateConnectionsForPrincipal(handler);
+	}
+
+	/**
+	 * Start to terminate connections applying the provided handler to the
+	 * principals of all connections.
+	 * 
+	 * Note: if {@link SessionCache} is used, it's not possible to remove a
+	 * cache entry, if no related connection is in the connection store.
+	 * 
+	 * @param principalHandler handler to be called within the serial execution
+	 *            of the related connection. If {@code true} is returned, the
+	 *            related connection is terminated
+	 * @return future to cancel or wait for completion
+	 */
+	public Future<Void> startTerminateConnectionsForPrincipal(
+			final LeastRecentlyUsedCache.Predicate<Principal> principalHandler) {
+		if (principalHandler == null) {
+			throw new NullPointerException("principal handler must not be null!");
+		}
+		LeastRecentlyUsedCache.Predicate<Connection> connectionHandler = new LeastRecentlyUsedCache.Predicate<Connection>() {
+
+			@Override
+			public boolean accept(Connection connection) {
+				Principal peer = null;
+				SessionTicket ticket = connection.getSessionTicket();
+				if (ticket != null) {
+					peer = ticket.getClientIdentity();
+				} else {
+					DTLSSession session = connection.getSession();
+					if (session != null) {
+						peer = session.getPeerIdentity();
+					}
+				}
+				if (peer != null && principalHandler.accept(peer)) {
+					connectionStore.remove(connection, true);
+				}
+				return false;
+			}
+		};
+		return startForEach(connectionHandler);
+	}
+
+	/**
+	 * Start applying provided handler to all connections.
+	 * 
+	 * @param handler handler to be called within the serial execution of the
+	 *            passed in connection. If {@code true} is returned, iterating
+	 *            is stopped.
+	 * @return future to cancel or wait for completion
+	 */
+	public Future<Void> startForEach(LeastRecentlyUsedCache.Predicate<Connection> handler) {
+		if (handler == null) {
+			throw new NullPointerException("handler must not be null!");
+		}
+		ForEachFuture result = new ForEachFuture();
+		nextForEach(connectionStore.iterator(), handler, result);
+		return result;
+	}
+
+	/**
+	 * Calls provided handler for each connection returned be the provided
+	 * iterator.
+	 * 
+	 * @param iterator iterator over connections
+	 * @param handler handler to be called for all connections returned by the
+	 *            iterator. Iteration is stopped, when handler returns
+	 *            {@code true}
+	 * @param result future to get cancelled or signal completion
+	 */
+	private void nextForEach(final Iterator<Connection> iterator,
+			final LeastRecentlyUsedCache.Predicate<Connection> handler, final ForEachFuture result) {
+		if (!result.isCancelled() && iterator.hasNext()) {
+			final Connection next = iterator.next();
+			try {
+				next.getExecutor().execute(new Runnable() {
+
+					@Override
+					public void run() {
```

Comment:
should we test result.isCancelled in this runnable too ?

Reply:
My assumption was, that `accept` will return very fast. If not, yes, we can check the cancelled.

Reply:
I would say "before to call the accept", just in case the runnable was blocked in serialexecutor.

:thinking:  but you mean this is not useful because `accept` will be fast and so we will go in `nextForEach` (which calls `isCancelled`)

I would agree that most of the time accept would be fast. Maybe sometime accept will do so store modification (deletion) and so it could make sense to test `isCancelled` before to avoid the modification ? 
(No strong opinion)

Reply:
> but you mean this is not useful because accept will be fast and so we will go in nextForEach (which calls isCancelled)

That's at least my assumption.

Cancel for such a thread-hoping execution will always be indeterministic.